### ISSUE TITLE
Eliminate vertex "stickiness" at the play area boundary

### ISSUE THREAD

**Issue by BLFiedler** [phetsims/quadrilateral#41]
We discussed trying to eliminate some of the "Stickiness" that occurs at the boundaries of the play area when dragging the vertices via cursor or touch.

We recognize that this also happens within the shape, but we will wait to address that if it seems problematic during interviews, etc

Vertices: Allow vertices to be dragged when the mouse leaves the play area, retaining only the coordinate along the axis of the play area boundary the cursor/touch crossed. e..g, if the cursor moves off of the left boundary, then allow the vertex to be dragged "up and down" along y-axis.

**Comment by jessegreenberg** [comment]
> We recognize that this also happens within the shape

This also happens with the constraint that no sides can overlap a vertex in #15.
![ezgif com-gif-maker (4)](https://user-images.githubusercontent.com/6396244/149980520-06c5bee2-3395-43e2-be25-ad10f2c0d7f3.gif)

**Comment by jessegreenberg** [comment]
This issue is not as easy as modelBounds.closestPointTo( mousePosition ) because we need to make sure that not just the mouse position is allowed but that the entire vertex bounds are contained in the drag area.  If we have to do some work for the stickiness along the boundary maybe it is worth thinking about a solution for anywhere in the drag area.

**Comment by jessegreenberg** [comment]
Discussed with @BLFiedler - It seems likely that we will want to solve this generally. This issue doesn't seem as important as some other tasks, @BLFiedler reduced the priority of this issue.

**Comment by BLFiedler** [comment]
Deferring this until we discuss issue #80

**Comment by jessegreenberg** [comment]
I tried a different approach where we use the intersection of lines along the x and y axes relative to the mouse position with the shape. But when both x and y axes intersect it is ambiguous which intersection point to use for the vertex. Here is an example of it switching between intersection points in a somewhat confusing way:
![rts](https://user-images.githubusercontent.com/6396244/165160692-b8a0b804-5db2-481d-8efa-fb19062c9196.gif)


Maybe https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1048918281 is still best but would be good to check in with @BLFiedler about this.

**Comment by jessegreenberg** [comment]
OK, I have a way to do this now using a Line function `explicitClosestToPoint`. Instead of creating the drag area shape we create `Line`s that outline the shape and iterate over each to find the closest point to the closest line. Then we use the collection of Lines to draw the drag area for each vertex.

Its working well along the boundary but the vertex/vertex and vertex/side collisions are not considered in the `Line`s and so the draggable Vertex still gets stuck around the interior shape.

**Comment by jessegreenberg** [comment]
Ideally we should modify what we did in https://github.com/phetsims/quadrilateral/issues/15 so that the single shape itself prevents vertex/side and vertex/vertex overlap.

**Comment by jessegreenberg** [comment]
Here are the (purple) points that define the shape currently:
![image](https://user-images.githubusercontent.com/6396244/165187069-186e8a5b-de42-4496-bd18-75581bfda486.png)

I think what we need is a shape like this
![image](https://user-images.githubusercontent.com/6396244/165187781-3c6bb540-74ec-4bc0-a480-6273f5feb26d.png)


The red lines indicate how we find the direction for the diagonal lines of the drag area shape to make sure that the right side doesn't overlap vertexA nd the top side doesn't overlap with verteC.

**Comment by jessegreenberg** [comment]
The red lines in https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1109118484 aren't quite right because they are dependent on the current position of the moving vertex. The shape for the moving vertex shouldn't change as it is dragged.

**Comment by jessegreenberg** [comment]
This morning I was able to get a portion of https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1109118484 but there is more to be done for more complicated and concave shapes.
![rts](https://user-images.githubusercontent.com/6396244/165326407-72780657-44b4-4f23-a630-1df64998e416.gif)

**Comment by jessegreenberg** [comment]
We discussed during meeting today:  Since we are close to using this solution, lets see it working for concave shapes if possible and then review again.

We also discussed that we could just erode the shape that is calculated in master. There is a question about whether that will work in all cases, but it may be simple and worth trying. In this approach @BLFiedler suggested that vertex/side overlap was OK  but it is most important to straight line creation with vertices and sides.

**Comment by jessegreenberg** [comment]
A pure scaling matrix applied to the shape isn't working out, I don't know how to reposition the shape control points after the scaling:

<details>

![image](https://user-images.githubusercontent.com/6396244/165370075-9324b2c3-1b9e-4c16-a508-ee8f86d10b4d.png)
</details>

I tried some hackiness using getStrokedShape to create a new shape from the control points from a line width. Its really close but Kite is wrapping around the points somehow. I tried reversing the control point order thinking there may be a clockwise/counter clocwise difference but no luck.

<details>

![image](https://user-images.githubusercontent.com/6396244/165370595-7776e569-cdd5-4f34-be21-3186cbb43c6a.png)

</details>

</details>

**Comment by jessegreenberg** [comment]
I have to move on for a bit but here is some progress:

<details>

```patch
Index: js/quadrilateral/view/VertexNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/VertexNode.ts b/js/quadrilateral/view/VertexNode.ts
--- a/js/quadrilateral/view/VertexNode.ts	(revision 269ea9d935bfe96820c1a6ec62207d14db212a2f)
+++ b/js/quadrilateral/view/VertexNode.ts	(date 1650992284261)
@@ -107,12 +107,19 @@
         const parentPoint = this.globalToParentPoint( pointerPoint! );
         const modelPoint = modelViewTransform.viewToModelPosition( parentPoint );
 
+        const closestDragAreaPoint = model.quadrilateralShapeModel.getClosestDragAreaPoint( vertex, modelPoint );
+
         // constrain to the allowable positions in the model along the grid
-        const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( modelPoint );
+        const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( closestDragAreaPoint );
 
-        if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
-          vertex.positionProperty.value = constrainedPosition;
-        }
+        vertex.positionProperty.value = constrainedPosition;
+
+        // @ts-ignore
+        // window.testCircle.center = modelViewTransform.modelToViewPosition( closestDragAreaPoint );
+
+        // if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
+        //   vertex.positionProperty.value = constrainedPosition;
+        // }
       },
 
       // @ts-ignore - TODO: Need to come through and do options, see #27
Index: js/quadrilateral/view/VertexDragAreaNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/VertexDragAreaNode.ts b/js/quadrilateral/view/VertexDragAreaNode.ts
--- a/js/quadrilateral/view/VertexDragAreaNode.ts	(revision 269ea9d935bfe96820c1a6ec62207d14db212a2f)
+++ b/js/quadrilateral/view/VertexDragAreaNode.ts	(date 1650996050285)
@@ -19,7 +19,8 @@
   constructor( vertex: Vertex, sides: Side[], modelViewTransform: ModelViewTransform2, providedOptions?: PathOptions ) {
 
     const options = optionize<PathOptions, {}>()( {
-      fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      // fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      fill: 'rgba(255,0,0,0.5)'
     }, providedOptions );
 
     super( null, options );
Index: js/quadrilateral/view/QuadrilateralScreenView.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/QuadrilateralScreenView.ts b/js/quadrilateral/view/QuadrilateralScreenView.ts
--- a/js/quadrilateral/view/QuadrilateralScreenView.ts	(revision 269ea9d935bfe96820c1a6ec62207d14db212a2f)
+++ b/js/quadrilateral/view/QuadrilateralScreenView.ts	(date 1650992284257)
@@ -12,7 +12,7 @@
 import QuadrilateralConstants from '../../common/QuadrilateralConstants.js';
 import quadrilateral from '../../quadrilateral.js';
 import QuadrilateralModel from '../model/QuadrilateralModel.js';
-import { Rectangle, Text } from '../../../../scenery/js/imports.js';
+import { Circle, Rectangle, Text } from '../../../../scenery/js/imports.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
 import QuadrilateralNode from './QuadrilateralNode.js';
 import QuadrilateralSoundView from './QuadrilateralSoundView.js';
@@ -243,6 +243,12 @@
       this.addChild( calibrationButton );
     }
 
+    const testCircle = new Circle( 10, { fill: 'blue' } );
+    this.addChild( testCircle );
+
+    // @ts-ignore
+    window.testCircle = testCircle;
+
     // pdom
     this.pdomPlayAreaNode.pdomOrder = [ this.quadrilateralNode ];
     this.pdomControlAreaNode.pdomOrder = [ visibilityControls, this.resetAllButton ];
Index: js/quadrilateral/model/Vertex.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/model/Vertex.ts b/js/quadrilateral/model/Vertex.ts
--- a/js/quadrilateral/model/Vertex.ts	(revision 269ea9d935bfe96820c1a6ec62207d14db212a2f)
+++ b/js/quadrilateral/model/Vertex.ts	(date 1651001796559)
@@ -14,7 +14,7 @@
 import Bounds2 from '../../../../dot/js/Bounds2.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
 import Vector2 from '../../../../dot/js/Vector2.js';
-import { Shape } from '../../../../kite/js/imports.js';
+import { Line, LineStyles, Shape } from '../../../../kite/js/imports.js';
 import IReadOnlyProperty from '../../../../axon/js/IReadOnlyProperty.js';
 import VertexLabel from './VertexLabel.js';
 
@@ -28,7 +28,10 @@
   public angleProperty: Property<number | null>;
   public readonly vertexLabel: VertexLabel;
   public dragBoundsProperty: Property<null | Bounds2>;
+
+  public dragAreaLines: Line[] = [];
   public dragAreaProperty: Property<null | Shape>;
+
   public isPressedProperty: Property<boolean>;
   public modelBoundsProperty: IReadOnlyProperty<Bounds2>;
   private tandem: Tandem;
@@ -99,6 +102,88 @@
     this.tandem = tandem;
   }
 
+  /**
+   * Set the lines that define the shape that constrains the dragging of this vertex.
+   */
+  public setDragAreaLines( lines: Line[] ): void {
+    this.dragAreaLines = lines;
+    this.dragAreaProperty.value = this.createDragAreaShapeFromLines( lines );
+  }
+
+  /**
+   * Create the shape that defines the drag area for this vertex by building it out of the
+   * drag area lines.
+   */
+  private createDragAreaShapeFromLines( lines: Line[] ): Shape {
+
+    // const points: Vector2[] = [];
+    // lines.forEach( line => {
+    //   points.push( line.start );
+    //   points.push( line.end );
+    // } );
+    // const polygon = Shape.polygon( points );
+    // const strokedPolygon = polygon.getStrokedShape( new LineStyles( { lineWidth: Vertex.VERTEX_WIDTH } ) );
+    // const innerPolygonPoints = strokedPolygon.subpaths[ 1 ].points;
+    //
+    // const innerLines: Line[] = [];
+    //
+    // for ( let i = 0; i < innerPolygonPoints.length - 1; i++ ) {
+    //   innerLines.push( new Line( innerPolygonPoints[ i ], innerPolygonPoints[ i + 1 ] ) );
+    // }
+    //
+    // // finish with the closing line after the loop
+    // innerLines.push( new Line( innerPolygonPoints[ innerPolygonPoints.length - 1 ], innerPolygonPoints[ 0 ] ) );
+    //
+    //
+    // const linesWithoutIntersections = [];
+    // for ( let i = 0; i < innerLines.length - 1; i++ ) {
+    //   debugger;
+    //   const intersection = Line.intersect( innerLines[ i ], innerLines[ i + 1 ] );
+    //   if ( intersection.length > 0 ) {
+    //     linesWithoutIntersections.push( new Line( innerLines[ i ].start, intersection[ 0 ].point ) );
+    //   }
+    //   else {
+    //     linesWithoutIntersections.push( innerLines[ i ] );
+    //   }
+    // }
+    //
+    // const shape = new Shape();
+    // if ( linesWithoutIntersections.length > 0 ) {
+    //   shape.moveToPoint( linesWithoutIntersections[ 0 ].start );
+    //
+    //   linesWithoutIntersections.forEach( line => {
+    //     shape.lineToPoint( line.end );
+    //   } );
+    //   shape.close();
+    // }
+    // return shape;
+
+    // TODO: Try getOffsetShape
+    // TODO: Continue with finding the intersections between lines to weed them out
+
+    // now loop over the inner path looking for inetrsection points between adjacent lines in the subpath.
+    // return strokedPolygon;
+    // const innerPolygonPoints = strokedPolygon.subpaths[ 1 ].points;
+    // console.log( points.length, innerPolygonPoints.length );
+    // const reversedInnerPolygonPoints = innerPolygonPoints.reverse();
+    // const innerPolygon = Shape.polygon( innerPolygonPoints );
+    // return innerPolygon;
+    // return polygon;
+
+    const shape = new Shape();
+
+    if ( lines.length > 0 ) {
+      shape.moveToPoint( lines[ 0 ].start );
+
+      lines.forEach( line => {
+        shape.lineToPoint( line.end );
+      } );
+      shape.close();
+    }
+
+    return shape.getOffsetShape( -0.03 );
+  }
+
   /**
    * Returns true when the provided bounds overlap the modelled bounds of this Vertex.
    */
Index: js/quadrilateral/model/QuadrilateralShapeModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/model/QuadrilateralShapeModel.ts b/js/quadrilateral/model/QuadrilateralShapeModel.ts
--- a/js/quadrilateral/model/QuadrilateralShapeModel.ts	(revision 269ea9d935bfe96820c1a6ec62207d14db212a2f)
+++ b/js/quadrilateral/model/QuadrilateralShapeModel.ts	(date 1650999709833)
@@ -18,7 +18,7 @@
 import Vertex from './Vertex.js';
 import Utils from '../../../../dot/js/Utils.js';
 import Emitter from '../../../../axon/js/Emitter.js';
-import { Line, Shape } from '../../../../kite/js/imports.js';
+import { ClosestToPointResult, Line, LineStyles, Shape } from '../../../../kite/js/imports.js';
 import QuadrilateralModel from './QuadrilateralModel.js';
 import NumberProperty from '../../../../axon/js/NumberProperty.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
@@ -34,6 +34,7 @@
 import VertexAngles from './VertexAngles.js';
 import ParallelSideChecker from './ParallelSideChecker.js';
 import IReadOnlyProperty from '../../../../axon/js/IReadOnlyProperty.js';
+import Matrix3 from '../../../../dot/js/Matrix3.js';
 
 // A useful type for calculations for the vertex Shapes which define where the Vertex can move depending on
 // the positions of the other vertices. Lines are along the bounds of model space and RayIntersections
@@ -596,7 +597,7 @@
    * @param vertexC - the next vertex from vertexB, moving clockwise
    * @param vertexD - the next vertex from vertexC, moving clockwise
    */
-  private createVertexArea( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Shape {
+  private createVertexDragAreaLines( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Line[] {
 
     const allVerticesInBounds = _.every( [ vertexA, vertexB, vertexC, vertexD ], vertex => modelBounds.containsPoint( vertex.positionProperty.value ) );
     const vertexPositionsUnique = _.uniqBy( [ vertexA, vertexB, vertexC, vertexD ].map( vertex => vertex.positionProperty.value.toString() ), positionString => {
@@ -611,7 +612,7 @@
 
       // The shape creation algorithm requires that all vertices are in bounds - we may need to handle this gracefully
       // so just return an empty shape in this case
-      return new Shape();
+      return [];
     }
 
     // Lines around the bounds to detect intersections - remember that for Bounds2 top and bottom
@@ -718,17 +719,39 @@
       points.push( vertexB.positionProperty.value ); // walk back to vertexB
     }
 
-    const shape = new Shape();
-    shape.moveToPoint( points[ 0 ] );
+    const lines = [];
+
+    if ( points.length > 0 ) {
+      for ( let i = 0; i < points.length - 1; i++ ) {
+        lines.push( new Line( points[ i ], points[ i + 1 ] ) );
+      }
 
-    for ( let i = 1; i < points.length; i++ ) {
-      shape.lineToPoint( points[ i ] );
+      // finish with the closing line after the loop
+      lines.push( new Line( points[ points.length - 1 ], points[ 0 ] ) );
     }
 
-    // closing the shape after the last intersection should bring us back to vertexC
-    shape.close();
+    return lines;
+
+    // return polygon.subpaths;
+
+    // turn the lines
+
+    // return lines;
+
+    // stroke each line to the right (since we are creating the lines in clockwise direction) to "erode" it to
+    // make it impossible for vertices and lines to perfectly overlap
+    // const erodedLines = lines.map( line => this.customStrokeRight( line, Vertex.VERTEX_WIDTH )[ 0 ] );
+    // return erodedLines;
+
+    // transorm the lines in a way that we do not allow vertex/vertex or vertex/side overlap
+    // const scalingMatrix = Matrix3.scaling( 0.9, 0.9 );
+    // const transformedLines = lines.map( line => line.transformed( scalingMatrix ) );
+    // return transformedLines;
+  }
 
-    return shape;
+  private customStrokeRight( line: Line, lineWidth: number ): Line[] {
+    const offset = line.getStartTangent().perpendicular.times( lineWidth / 2 );
+    return [ new Line( line.start.plus( offset ), line.end.plus( offset ) ) ];
   }
 
   /**
@@ -742,7 +765,8 @@
     const lengthA = line.start.distance( point );
     const lengthB = line.end.distance( point );
 
-    return length === lengthA + lengthB;
+    // TODO; This is fragile, is there a better way? Precision errors may break this equality check and then we don't calculate a boundary intersection point correctly.
+    return Utils.equalsEpsilon( length, lengthA + lengthB, 0.00001 );
   }
 
   /**
@@ -872,6 +896,68 @@
     return shapeAllowed;
   }
 
+  /**
+   * Returns the closest point to the drag area of a vertex from the provided point. If provided point is inside
+   * the drag area, the point is returned.
+   *
+   * @param vertex - The vertex to move (TODO: Should this function be moved to Vertex.ts)
+   * @param point - in the model coordinate frame
+   */
+  getClosestDragAreaPoint( vertex: Vertex, point: Vector2 ) {
+
+    if ( !vertex.dragAreaProperty.value!.containsPoint( point ) ) {
+      const results: ClosestToPointResult[] = [];
+      vertex.dragAreaLines.forEach( dragAreaLine => {
+        results.push( ...dragAreaLine.explicitClosestToPoint( point ) );
+      } );
+
+      const closestResult = _.minBy( results, result => result.distanceSquared );
+      assert && assert( closestResult, 'Why wasnt there a result' );
+      return closestResult!.closestPoint;
+    }
+    else {
+
+      // If we are inside the drag area, just return the proposed point
+      return point;
+    }
+
+
+    // // ray from the vertex position to the mouse point
+    // const distance = point.minus( vertex.positionProperty.value );
+    // if ( distance.magnitude !== 0 ) {
+    //
+    //   // const horizontalDirection = distance.x > 0 ? Vector2.X_UNIT.negated() : Vector2.X_UNIT;
+    //   // const horizontalRay = new Ray2( point, horizontalDirection );
+    //   //
+    //   // const verticalDirection = distance.y > 0 ? Vector2.Y_UNIT.negated() : Vector2.Y_UNIT;
+    //   // const verticalRay = new Ray2( point, verticalDirection );
+    //   //
+    //   const directDirection = vertex.positionProperty.value.minus( point ).normalized();
+    //   const directRay = new Ray2( point, directDirection );
+    //
+    //   // // get the intersection points
+    //   // const horizontalIntersections = vertex.dragAreaProperty.value!.intersection( horizontalRay );
+    //   // const verticalIntersections = vertex.dragAreaProperty.value!.intersection( verticalRay );
+    //   //
+    //   // if ( horizontalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( horizontalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.x = closestIntersection.point.x;
+    //   // }
+    //   // else if ( verticalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( verticalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.y = closestIntersection.point.y;
+    //   // }
+    //
+    //   const directIntersections = vertex.dragAreaProperty.value!.intersection( directRay );
+    //   const closestIntersection = _.minBy( directIntersections, intersection => intersection.point.distance( point ) );
+    //   closestPoint.set( closestIntersection.point );
+    //
+    //   console.log( closestPoint, point );
+    // }
+
+    // return closestPoint;
+  }
+
   /**
    * Returns whether or not the quadrilateral shape is a parallelogram, within the tolerance defined by
    * angleToleranceIntervalProperty. This function uses parallelSidePairsProperty and requires that Property
@@ -951,7 +1037,6 @@
     } );
 
     // un-defer all so that all Properties and calls callbacks
-    // debugger;
     this.setPropertiesDeferred( false );
   }
 
@@ -1272,10 +1357,10 @@
    * Update the drag areas for all vertices.
    */
   private setVertexDragAreas(): void {
-    this.vertexA.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
-    this.vertexB.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
-    this.vertexC.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
-    this.vertexD.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
+    this.vertexA.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
+    this.vertexB.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
+    this.vertexC.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
+    this.vertexD.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
   }
 
   public setPropertiesDeferred( deferred: boolean ): void {

```
</details>

**Comment by jessegreenberg** [comment]
I tried using `getOffsetShape` but it looks the same as `getStrokedShape`. I am going to bail on this it seems like it is no easier than manually calculating points.

**Comment by BLFiedler** [comment]
Yeah, seems like it's own rabbit hole. Perhaps returning to solving for the concave case in https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1109884900 is better?

**Comment by jessegreenberg** [comment]
Yes I decided to move back to that approach at the end of https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1110135452. The problem in https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1111083130 is present in the proposal from the original approach from https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1109884900.

**Comment by jessegreenberg** [comment]
I tried a few more things here:
- Getting the centroid of the shape before scaling it down, then scaling it down, then translating it back to the original centroid
- Getting the centroid of the shape points and pulling all points toward the centroid to reduce the area

They worked OK for lots of cases but broke in many others.

Here is a patch

<details>

```patch
Index: quadrilateral/js/quadrilateral/view/VertexNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/quadrilateral/js/quadrilateral/view/VertexNode.ts b/quadrilateral/js/quadrilateral/view/VertexNode.ts
--- a/quadrilateral/js/quadrilateral/view/VertexNode.ts	(revision 9b8d249cbee6cb3c6c4e8ad1356c7be4dde99cba)
+++ b/quadrilateral/js/quadrilateral/view/VertexNode.ts	(date 1651075846464)
@@ -107,12 +107,19 @@
         const parentPoint = this.globalToParentPoint( pointerPoint! );
         const modelPoint = modelViewTransform.viewToModelPosition( parentPoint );
 
+        const closestDragAreaPoint = model.quadrilateralShapeModel.getClosestDragAreaPoint( vertex, modelPoint );
+
         // constrain to the allowable positions in the model along the grid
-        const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( modelPoint );
+        const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( closestDragAreaPoint );
 
-        if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
-          vertex.positionProperty.value = constrainedPosition;
-        }
+        vertex.positionProperty.value = constrainedPosition;
+
+        // @ts-ignore
+        // window.testCircle.center = modelViewTransform.modelToViewPosition( closestDragAreaPoint );
+
+        // if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
+        //   vertex.positionProperty.value = constrainedPosition;
+        // }
       },
 
       // @ts-ignore - TODO: Need to come through and do options, see #27
Index: quadrilateral/js/quadrilateral/view/VertexDragAreaNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/quadrilateral/js/quadrilateral/view/VertexDragAreaNode.ts b/quadrilateral/js/quadrilateral/view/VertexDragAreaNode.ts
--- a/quadrilateral/js/quadrilateral/view/VertexDragAreaNode.ts	(revision 9b8d249cbee6cb3c6c4e8ad1356c7be4dde99cba)
+++ b/quadrilateral/js/quadrilateral/view/VertexDragAreaNode.ts	(date 1651075846460)
@@ -19,7 +19,8 @@
   constructor( vertex: Vertex, sides: Side[], modelViewTransform: ModelViewTransform2, providedOptions?: PathOptions ) {
 
     const options = optionize<PathOptions, {}>()( {
-      fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      // fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      fill: 'rgba(255,0,0,0.5)'
     }, providedOptions );
 
     super( null, options );
Index: quadrilateral/js/quadrilateral/view/QuadrilateralScreenView.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/quadrilateral/js/quadrilateral/view/QuadrilateralScreenView.ts b/quadrilateral/js/quadrilateral/view/QuadrilateralScreenView.ts
--- a/quadrilateral/js/quadrilateral/view/QuadrilateralScreenView.ts	(revision 9b8d249cbee6cb3c6c4e8ad1356c7be4dde99cba)
+++ b/quadrilateral/js/quadrilateral/view/QuadrilateralScreenView.ts	(date 1651075846457)
@@ -12,7 +12,7 @@
 import QuadrilateralConstants from '../../common/QuadrilateralConstants.js';
 import quadrilateral from '../../quadrilateral.js';
 import QuadrilateralModel from '../model/QuadrilateralModel.js';
-import { Rectangle, Text } from '../../../../scenery/js/imports.js';
+import { Circle, Rectangle, Text } from '../../../../scenery/js/imports.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
 import QuadrilateralNode from './QuadrilateralNode.js';
 import QuadrilateralSoundView from './QuadrilateralSoundView.js';
@@ -243,6 +243,12 @@
       this.addChild( calibrationButton );
     }
 
+    const testCircle = new Circle( 10, { fill: 'blue' } );
+    this.addChild( testCircle );
+
+    // @ts-ignore
+    window.testCircle = testCircle;
+
     // pdom
     this.pdomPlayAreaNode.pdomOrder = [ this.quadrilateralNode ];
     this.pdomControlAreaNode.pdomOrder = [ visibilityControls, this.resetAllButton ];
Index: quadrilateral/js/quadrilateral/model/Vertex.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/quadrilateral/js/quadrilateral/model/Vertex.ts b/quadrilateral/js/quadrilateral/model/Vertex.ts
--- a/quadrilateral/js/quadrilateral/model/Vertex.ts	(revision 9b8d249cbee6cb3c6c4e8ad1356c7be4dde99cba)
+++ b/quadrilateral/js/quadrilateral/model/Vertex.ts	(date 1651078030554)
@@ -14,7 +14,7 @@
 import Bounds2 from '../../../../dot/js/Bounds2.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
 import Vector2 from '../../../../dot/js/Vector2.js';
-import { Shape } from '../../../../kite/js/imports.js';
+import { Line, Shape } from '../../../../kite/js/imports.js';
 import IReadOnlyProperty from '../../../../axon/js/IReadOnlyProperty.js';
 import VertexLabel from './VertexLabel.js';
 
@@ -28,7 +28,10 @@
   public angleProperty: Property<number | null>;
   public readonly vertexLabel: VertexLabel;
   public dragBoundsProperty: Property<null | Bounds2>;
+
+  public dragAreaLines: Line[] = [];
   public dragAreaProperty: Property<null | Shape>;
+
   public isPressedProperty: Property<boolean>;
   public modelBoundsProperty: IReadOnlyProperty<Bounds2>;
   private tandem: Tandem;
@@ -99,6 +102,107 @@
     this.tandem = tandem;
   }
 
+  /**
+   * Set the lines that define the shape that constrains the dragging of this vertex.
+   */
+  public setDragAreaLines( lines: Line[] ): void {
+    this.dragAreaLines = lines;
+    this.dragAreaProperty.value = this.createDragAreaShapeFromLines( lines );
+  }
+
+  /**
+   * Create the shape that defines the drag area for this vertex by building it out of the
+   * drag area lines.
+   */
+  private createDragAreaShapeFromLines( lines: Line[] ): Shape {
+
+    const points: Vector2[] = [];
+    lines.forEach( line => {
+      points.push( line.start );
+      points.push( line.end );
+    } );
+    // const polygon = Shape.polygon( points );
+    // const strokedPolygon = polygon.getStrokedShape( new LineStyles( { lineWidth: Vertex.VERTEX_WIDTH } ) );
+    // const innerPolygonPoints = strokedPolygon.subpaths[ 0 ].points;
+    // return Shape.polygon( innerPolygonPoints );
+
+    // const originalCentroid = Vector2.average( points );
+    // const originalCentroid = Vector2.ZERO;
+    const originalCentroid = this.positionProperty.value;
+    if ( this.vertexLabel === VertexLabel.VERTEX_A ) {
+      console.log( originalCentroid );
+    }
+
+    // pull all points toward the centroid
+    const pulledPoints = points.map( point => {
+      const deltaVector = originalCentroid.minus( point ).withMagnitude( Vertex.VERTEX_WIDTH );
+      return point.plus( deltaVector );
+    } );
+
+
+    // return Shape.polygon( points );
+    return Shape.polygon( pulledPoints );
+
+
+    // const innerLines: Line[] = [];
+    //
+    // for ( let i = 0; i < innerPolygonPoints.length - 1; i++ ) {
+    //   innerLines.push( new Line( innerPolygonPoints[ i ], innerPolygonPoints[ i + 1 ] ) );
+    // }
+    //
+    // // finish with the closing line after the loop
+    // innerLines.push( new Line( innerPolygonPoints[ innerPolygonPoints.length - 1 ], innerPolygonPoints[ 0 ] ) );
+    //
+    //
+    // const linesWithoutIntersections = [];
+    // for ( let i = 0; i < innerLines.length - 1; i++ ) {
+    //   debugger;
+    //   const intersection = Line.intersect( innerLines[ i ], innerLines[ i + 1 ] );
+    //   if ( intersection.length > 0 ) {
+    //     linesWithoutIntersections.push( new Line( innerLines[ i ].start, intersection[ 0 ].point ) );
+    //   }
+    //   else {
+    //     linesWithoutIntersections.push( innerLines[ i ] );
+    //   }
+    // }
+    //
+    // const shape = new Shape();
+    // if ( linesWithoutIntersections.length > 0 ) {
+    //   shape.moveToPoint( linesWithoutIntersections[ 0 ].start );
+    //
+    //   linesWithoutIntersections.forEach( line => {
+    //     shape.lineToPoint( line.end );
+    //   } );
+    //   shape.close();
+    // }
+    // return shape;
+
+    // TODO: Try getOffsetShape
+    // TODO: Continue with finding the intersections between lines to weed them out
+
+    // now loop over the inner path looking for inetrsection points between adjacent lines in the subpath.
+    // return strokedPolygon;
+    // const innerPolygonPoints = strokedPolygon.subpaths[ 1 ].points;
+    // console.log( points.length, innerPolygonPoints.length );
+    // const reversedInnerPolygonPoints = innerPolygonPoints.reverse();
+    // const innerPolygon = Shape.polygon( innerPolygonPoints );
+    // return innerPolygon;
+    // return polygon;
+
+    // const shape = new Shape();
+    //
+    // if ( lines.length > 0 ) {
+    //   shape.moveToPoint( lines[ 0 ].start );
+    //
+    //   lines.forEach( line => {
+    //     shape.lineToPoint( line.end );
+    //   } );
+    //   shape.close();
+    // }
+    //
+    // return shape.getOffsetShape( -0.03 );
+  }
+
   /**
    * Returns true when the provided bounds overlap the modelled bounds of this Vertex.
    */
Index: quadrilateral/js/quadrilateral/model/QuadrilateralShapeModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/quadrilateral/js/quadrilateral/model/QuadrilateralShapeModel.ts b/quadrilateral/js/quadrilateral/model/QuadrilateralShapeModel.ts
--- a/quadrilateral/js/quadrilateral/model/QuadrilateralShapeModel.ts	(revision 9b8d249cbee6cb3c6c4e8ad1356c7be4dde99cba)
+++ b/quadrilateral/js/quadrilateral/model/QuadrilateralShapeModel.ts	(date 1651075846433)
@@ -18,7 +18,7 @@
 import Vertex from './Vertex.js';
 import Utils from '../../../../dot/js/Utils.js';
 import Emitter from '../../../../axon/js/Emitter.js';
-import { Line, Shape } from '../../../../kite/js/imports.js';
+import { ClosestToPointResult, Line, LineStyles, Shape } from '../../../../kite/js/imports.js';
 import QuadrilateralModel from './QuadrilateralModel.js';
 import NumberProperty from '../../../../axon/js/NumberProperty.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
@@ -34,6 +34,7 @@
 import VertexAngles from './VertexAngles.js';
 import ParallelSideChecker from './ParallelSideChecker.js';
 import IReadOnlyProperty from '../../../../axon/js/IReadOnlyProperty.js';
+import Matrix3 from '../../../../dot/js/Matrix3.js';
 
 // A useful type for calculations for the vertex Shapes which define where the Vertex can move depending on
 // the positions of the other vertices. Lines are along the bounds of model space and RayIntersections
@@ -596,7 +597,7 @@
    * @param vertexC - the next vertex from vertexB, moving clockwise
    * @param vertexD - the next vertex from vertexC, moving clockwise
    */
-  private createVertexArea( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Shape {
+  private createVertexDragAreaLines( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Line[] {
 
     const allVerticesInBounds = _.every( [ vertexA, vertexB, vertexC, vertexD ], vertex => modelBounds.containsPoint( vertex.positionProperty.value ) );
     const vertexPositionsUnique = _.uniqBy( [ vertexA, vertexB, vertexC, vertexD ].map( vertex => vertex.positionProperty.value.toString() ), positionString => {
@@ -611,7 +612,7 @@
 
       // The shape creation algorithm requires that all vertices are in bounds - we may need to handle this gracefully
       // so just return an empty shape in this case
-      return new Shape();
+      return [];
     }
 
     // Lines around the bounds to detect intersections - remember that for Bounds2 top and bottom
@@ -718,17 +719,39 @@
       points.push( vertexB.positionProperty.value ); // walk back to vertexB
     }
 
-    const shape = new Shape();
-    shape.moveToPoint( points[ 0 ] );
+    const lines = [];
+
+    if ( points.length > 0 ) {
+      for ( let i = 0; i < points.length - 1; i++ ) {
+        lines.push( new Line( points[ i ], points[ i + 1 ] ) );
+      }
 
-    for ( let i = 1; i < points.length; i++ ) {
-      shape.lineToPoint( points[ i ] );
+      // finish with the closing line after the loop
+      lines.push( new Line( points[ points.length - 1 ], points[ 0 ] ) );
     }
 
-    // closing the shape after the last intersection should bring us back to vertexC
-    shape.close();
+    return lines;
+
+    // return polygon.subpaths;
+
+    // turn the lines
+
+    // return lines;
+
+    // stroke each line to the right (since we are creating the lines in clockwise direction) to "erode" it to
+    // make it impossible for vertices and lines to perfectly overlap
+    // const erodedLines = lines.map( line => this.customStrokeRight( line, Vertex.VERTEX_WIDTH )[ 0 ] );
+    // return erodedLines;
+
+    // transorm the lines in a way that we do not allow vertex/vertex or vertex/side overlap
+    // const scalingMatrix = Matrix3.scaling( 0.9, 0.9 );
+    // const transformedLines = lines.map( line => line.transformed( scalingMatrix ) );
+    // return transformedLines;
+  }
 
-    return shape;
+  private customStrokeRight( line: Line, lineWidth: number ): Line[] {
+    const offset = line.getStartTangent().perpendicular.times( lineWidth / 2 );
+    return [ new Line( line.start.plus( offset ), line.end.plus( offset ) ) ];
   }
 
   /**
@@ -742,7 +765,8 @@
     const lengthA = line.start.distance( point );
     const lengthB = line.end.distance( point );
 
-    return length === lengthA + lengthB;
+    // TODO; This is fragile, is there a better way? Precision errors may break this equality check and then we don't calculate a boundary intersection point correctly.
+    return Utils.equalsEpsilon( length, lengthA + lengthB, 0.00001 );
   }
 
   /**
@@ -872,6 +896,68 @@
     return shapeAllowed;
   }
 
+  /**
+   * Returns the closest point to the drag area of a vertex from the provided point. If provided point is inside
+   * the drag area, the point is returned.
+   *
+   * @param vertex - The vertex to move (TODO: Should this function be moved to Vertex.ts)
+   * @param point - in the model coordinate frame
+   */
+  getClosestDragAreaPoint( vertex: Vertex, point: Vector2 ) {
+
+    if ( !vertex.dragAreaProperty.value!.containsPoint( point ) ) {
+      const results: ClosestToPointResult[] = [];
+      vertex.dragAreaLines.forEach( dragAreaLine => {
+        results.push( ...dragAreaLine.explicitClosestToPoint( point ) );
+      } );
+
+      const closestResult = _.minBy( results, result => result.distanceSquared );
+      assert && assert( closestResult, 'Why wasnt there a result' );
+      return closestResult!.closestPoint;
+    }
+    else {
+
+      // If we are inside the drag area, just return the proposed point
+      return point;
+    }
+
+
+    // // ray from the vertex position to the mouse point
+    // const distance = point.minus( vertex.positionProperty.value );
+    // if ( distance.magnitude !== 0 ) {
+    //
+    //   // const horizontalDirection = distance.x > 0 ? Vector2.X_UNIT.negated() : Vector2.X_UNIT;
+    //   // const horizontalRay = new Ray2( point, horizontalDirection );
+    //   //
+    //   // const verticalDirection = distance.y > 0 ? Vector2.Y_UNIT.negated() : Vector2.Y_UNIT;
+    //   // const verticalRay = new Ray2( point, verticalDirection );
+    //   //
+    //   const directDirection = vertex.positionProperty.value.minus( point ).normalized();
+    //   const directRay = new Ray2( point, directDirection );
+    //
+    //   // // get the intersection points
+    //   // const horizontalIntersections = vertex.dragAreaProperty.value!.intersection( horizontalRay );
+    //   // const verticalIntersections = vertex.dragAreaProperty.value!.intersection( verticalRay );
+    //   //
+    //   // if ( horizontalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( horizontalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.x = closestIntersection.point.x;
+    //   // }
+    //   // else if ( verticalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( verticalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.y = closestIntersection.point.y;
+    //   // }
+    //
+    //   const directIntersections = vertex.dragAreaProperty.value!.intersection( directRay );
+    //   const closestIntersection = _.minBy( directIntersections, intersection => intersection.point.distance( point ) );
+    //   closestPoint.set( closestIntersection.point );
+    //
+    //   console.log( closestPoint, point );
+    // }
+
+    // return closestPoint;
+  }
+
   /**
    * Returns whether or not the quadrilateral shape is a parallelogram, within the tolerance defined by
    * angleToleranceIntervalProperty. This function uses parallelSidePairsProperty and requires that Property
@@ -951,7 +1037,6 @@
     } );
 
     // un-defer all so that all Properties and calls callbacks
-    // debugger;
     this.setPropertiesDeferred( false );
   }
 
@@ -1272,10 +1357,10 @@
    * Update the drag areas for all vertices.
    */
   private setVertexDragAreas(): void {
-    this.vertexA.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
-    this.vertexB.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
-    this.vertexC.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
-    this.vertexD.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
+    this.vertexA.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
+    this.vertexB.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
+    this.vertexC.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
+    this.vertexD.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
   }
 
   public setPropertiesDeferred( deferred: boolean ): void {

```
</details>

I am stumped and I think I should move on to other issues for quad. But one thought I had to move forward is to use the shape we already have from https://github.com/phetsims/quadrilateral/issues/15 and not use the shape to prevent vertex/vertex and vertex/side overlap. Maybe there is a different way we can prevent that from happening and still allow smooth dragging.

**Comment by BLFiedler** [comment]
Right -  I agree with moving over to Voicing issues for now.

We can meet to discuss on Monday or so if it'd be helpful to think about what might work best here. Maybe revisiting the angle restrictions? I think I am more concerned with vertices overlapping and angles near 0 (forming lines) than I am some portion of the sides overlapping the vertices (or rather being under them).

**Comment by jessegreenberg** [comment]
I had some success with implementing this in a way that is demonstratable, we can play with it in this patch:

<details>

```patch
Index: js/quadrilateral/view/VertexNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/VertexNode.ts b/js/quadrilateral/view/VertexNode.ts
--- a/js/quadrilateral/view/VertexNode.ts	(revision 933cf76d8cb8303637c591452e3008dfdbd36876)
+++ b/js/quadrilateral/view/VertexNode.ts	(date 1651245273551)
@@ -121,12 +121,19 @@
         const parentPoint = this.globalToParentPoint( pointerPoint! );
         const modelPoint = modelViewTransform.viewToModelPosition( parentPoint );
 
+        const closestDragAreaPoint = model.quadrilateralShapeModel.getClosestDragAreaPoint( vertex, modelPoint );
         // constrain to the allowable positions in the model along the grid
-        const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( modelPoint );
+
+        // const constrainedPosition = QuadrilateralModel.getClosestMinorGridPosition( closestDragAreaPoint );
 
-        if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
-          vertex.positionProperty.value = constrainedPosition;
-        }
+        vertex.positionProperty.value = closestDragAreaPoint;
+
+        // @ts-ignore
+        // window.testCircle.center = modelViewTransform.modelToViewPosition( closestDragAreaPoint );
+
+        // if ( model.isVertexPositionAllowed( vertex, constrainedPosition ) ) {
+        //   vertex.positionProperty.value = constrainedPosition;
+        // }
       },
 
       // @ts-ignore - TODO: Need to come through and do options, see #27
Index: js/quadrilateral/view/VertexDragAreaNode.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/VertexDragAreaNode.ts b/js/quadrilateral/view/VertexDragAreaNode.ts
--- a/js/quadrilateral/view/VertexDragAreaNode.ts	(revision 933cf76d8cb8303637c591452e3008dfdbd36876)
+++ b/js/quadrilateral/view/VertexDragAreaNode.ts	(date 1651246280673)
@@ -19,7 +19,8 @@
   constructor( vertex: Vertex, sides: Side[], modelViewTransform: ModelViewTransform2, providedOptions?: PathOptions ) {
 
     const options = optionize<PathOptions, {}>()( {
-      fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      // fill: `rgba(${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},${dotRandom.nextInt( 255 )},0.5)`
+      fill: 'rgba(255,0,0,0.5)'
     }, providedOptions );
 
     super( null, options );
Index: js/quadrilateral/view/QuadrilateralScreenView.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/view/QuadrilateralScreenView.ts b/js/quadrilateral/view/QuadrilateralScreenView.ts
--- a/js/quadrilateral/view/QuadrilateralScreenView.ts	(revision 933cf76d8cb8303637c591452e3008dfdbd36876)
+++ b/js/quadrilateral/view/QuadrilateralScreenView.ts	(date 1651245273545)
@@ -12,7 +12,7 @@
 import QuadrilateralConstants from '../../common/QuadrilateralConstants.js';
 import quadrilateral from '../../quadrilateral.js';
 import QuadrilateralModel from '../model/QuadrilateralModel.js';
-import { Rectangle, Text } from '../../../../scenery/js/imports.js';
+import { Circle, Rectangle, Text } from '../../../../scenery/js/imports.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
 import QuadrilateralNode from './QuadrilateralNode.js';
 import QuadrilateralSoundView from './QuadrilateralSoundView.js';
@@ -242,6 +242,12 @@
       this.addChild( calibrationButton );
     }
 
+    const testCircle = new Circle( 10, { fill: 'blue' } );
+    this.addChild( testCircle );
+
+    // @ts-ignore
+    window.testCircle = testCircle;
+
     // pdom
     this.pdomPlayAreaNode.pdomOrder = [ this.quadrilateralNode ];
     this.pdomControlAreaNode.pdomOrder = [ visibilityControls, this.resetAllButton ];
Index: js/quadrilateral/model/Vertex.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/model/Vertex.ts b/js/quadrilateral/model/Vertex.ts
--- a/js/quadrilateral/model/Vertex.ts	(revision 933cf76d8cb8303637c591452e3008dfdbd36876)
+++ b/js/quadrilateral/model/Vertex.ts	(date 1651251573870)
@@ -14,7 +14,7 @@
 import Bounds2 from '../../../../dot/js/Bounds2.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
 import Vector2 from '../../../../dot/js/Vector2.js';
-import { Shape } from '../../../../kite/js/imports.js';
+import { Line, Shape } from '../../../../kite/js/imports.js';
 import IReadOnlyProperty from '../../../../axon/js/IReadOnlyProperty.js';
 import VertexLabel from './VertexLabel.js';
 
@@ -28,7 +28,10 @@
   public angleProperty: Property<number | null>;
   public readonly vertexLabel: VertexLabel;
   public dragBoundsProperty: Property<null | Bounds2>;
+
+  public dragAreaLines: Line[] = [];
   public dragAreaProperty: Property<null | Shape>;
+
   public isPressedProperty: Property<boolean>;
   public modelBoundsProperty: IReadOnlyProperty<Bounds2>;
   private tandem: Tandem;
@@ -99,6 +102,39 @@
     this.tandem = tandem;
   }
 
+  /**
+   * Set the lines that define the shape that constrains the dragging of this vertex.
+   */
+  public setDragAreaLines( lines: Line[] ): void {
+    this.dragAreaLines = lines;
+    this.dragAreaProperty.value = this.createDragAreaShapeFromLines( lines );
+  }
+
+  /**
+   * Create the shape that defines the drag area for this vertex by building it out of the
+   * drag area lines.
+   */
+  private createDragAreaShapeFromLines( lines: Line[] ): Shape {
+
+    const points: Vector2[] = [];
+    lines.forEach( line => {
+      points.push( line.start );
+      points.push( line.end );
+    } );
+
+    const shape = new Shape();
+
+    // shape.moveToPoint( points[ 0 ] );
+    points.forEach( point => {
+      // shape.circle( point.x, point.y, 0.01 ).newSubpath();
+      shape.lineToPoint( point );
+    } );
+
+    shape.close();
+
+    return shape;
+  }
+
   /**
    * Returns true when the provided bounds overlap the modelled bounds of this Vertex.
    */
Index: js/quadrilateral/model/QuadrilateralShapeModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/quadrilateral/model/QuadrilateralShapeModel.ts b/js/quadrilateral/model/QuadrilateralShapeModel.ts
--- a/js/quadrilateral/model/QuadrilateralShapeModel.ts	(revision 933cf76d8cb8303637c591452e3008dfdbd36876)
+++ b/js/quadrilateral/model/QuadrilateralShapeModel.ts	(date 1651251098077)
@@ -18,7 +18,7 @@
 import Vertex from './Vertex.js';
 import Utils from '../../../../dot/js/Utils.js';
 import Emitter from '../../../../axon/js/Emitter.js';
-import { Line, Shape } from '../../../../kite/js/imports.js';
+import { ClosestToPointResult, Line } from '../../../../kite/js/imports.js';
 import QuadrilateralModel from './QuadrilateralModel.js';
 import NumberProperty from '../../../../axon/js/NumberProperty.js';
 import QuadrilateralQueryParameters from '../QuadrilateralQueryParameters.js';
@@ -635,7 +635,7 @@
    * @param vertexC - the next vertex from vertexB, moving clockwise
    * @param vertexD - the next vertex from vertexC, moving clockwise
    */
-  private createVertexArea( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Shape {
+  private createVertexDragAreaLines( modelBounds: Bounds2, vertexA: Vertex, vertexB: Vertex, vertexC: Vertex, vertexD: Vertex ): Line[] {
 
     const allVerticesInBounds = _.every( [ vertexA, vertexB, vertexC, vertexD ], vertex => modelBounds.containsPoint( vertex.positionProperty.value ) );
     const vertexPositionsUnique = _.uniqBy( [ vertexA, vertexB, vertexC, vertexD ].map( vertex => vertex.positionProperty.value.toString() ), positionString => {
@@ -650,7 +650,7 @@
 
       // The shape creation algorithm requires that all vertices are in bounds - we may need to handle this gracefully
       // so just return an empty shape in this case
-      return new Shape();
+      return [];
     }
 
     // Lines around the bounds to detect intersections - remember that for Bounds2 top and bottom
@@ -758,17 +758,68 @@
       points.push( vertexB.positionProperty.value ); // walk back to vertexB
     }
 
-    const shape = new Shape();
-    shape.moveToPoint( points[ 0 ] );
+    const lines = [];
+
+    if ( points.length > 0 ) {
+      // assert && assert( points.length % 2 === 0,
+      //   'if we have points there should be an even number to form shape lines' );
+      for ( let i = 0; i < points.length - 1; i++ ) {
+        lines.push( new Line( points[ i ]!, points[ i + 1 ]! ) );
+      }
+
+      // finish with the closing line after the loop
+      lines.push( new Line( points[ points.length - 1 ]!, points[ 0 ]! ) );
+    }
+
+    // reduce the available area by the width of the line so that it is impossible for lines to form a zero
+    // angle
+    const reducedLines = lines.map( line => {
+      const strokedLine = line.strokeRight( Vertex.VERTEX_WIDTH * 2 )[ 0 ];
+      return new Line( strokedLine.end, strokedLine.start );
+    } );
+
+    const prunedLines: Line[] = [];
+    const addToPrunedLines = ( thisLine: Line, nextLine: Line, nextStart: Vector2 ): Vector2 => {
+      const intersections = Line.intersectOther( thisLine, nextLine );
 
-    for ( let i = 1; i < points.length; i++ ) {
-      shape.lineToPoint( points[ i ] );
-    }
+      if ( intersections.length > 0 ) {
+        const intersectionPoint = intersections[ 0 ].point;
+        prunedLines.push( new Line( nextStart, intersectionPoint ) );
+        nextStart = intersectionPoint;
+      }
+      else {
+        prunedLines.push( thisLine );
+        nextStart = thisLine.end;
+      }
 
-    // closing the shape after the last intersection should bring us back to vertexC
-    shape.close();
+      return nextStart;
+    };
 
-    return shape;
+    let nextStart: Vector2 = reducedLines[ 0 ].start;
+    for ( let i = 0; i < reducedLines.length - 1; i++ ) {
+
+      const thisLine = reducedLines[ i ];
+      const nextLine = reducedLines[ i + 1 ];
+      nextStart = addToPrunedLines( thisLine, nextLine, nextStart );
+    }
+
+    // Close the loop by returning to the first line
+    addToPrunedLines( reducedLines[ reducedLines.length - 1 ], reducedLines[ 0 ], nextStart );
+
+    // if ( vertexA.vertexLabel === VertexLabel.VERTEX_B ) {
+    //   debugger;
+    // }
+
+    prunedLines[ 0 ].start = prunedLines[ prunedLines.length - 1 ].end;
+
+    return prunedLines;
+  }
+
+  /**
+   * A unique implementation of positionAt that allows the parameteric position to extend beyond 0 to 1.
+   */
+  private static customPositionAt( line: Line, t: number ): Vector2 {
+    return line.start.plus( line.end.minus( line.start ).times( t ) );
   }
 
   /**
@@ -913,6 +964,71 @@
     return shapeAllowed;
   }
 
+  /**
+   * Returns the closest point to the drag area of a vertex from the provided point. If provided point is inside
+   * the drag area, the point is returned.
+   *
+   * @param vertex - The vertex to move (TODO: Should this function be moved to Vertex.ts)
+   * @param point - in the model coordinate frame
+   */
+  getClosestDragAreaPoint( vertex: Vertex, point: Vector2 ) {
+
+    if ( !vertex.dragAreaProperty.value!.containsPoint( point ) ) {
+      const results: ClosestToPointResult[] = [];
+      vertex.dragAreaLines.forEach( dragAreaLine => {
+        // Ignore lines with zero length which may be around due to the calculation of the dragAreaLines
+        if ( dragAreaLine.getArcLength() > 0 ) {
+          results.push( ...dragAreaLine.explicitClosestToPoint( point ) );
+        }
+      } );
+
+      const closestResult = _.minBy( results, result => result.distanceSquared );
+      assert && assert( closestResult, 'Why wasnt there a result' );
+      return closestResult!.closestPoint;
+    }
+    else {
+
+      // If we are inside the drag area, just return the proposed point
+      return point;
+    }
+
+
+    // // ray from the vertex position to the mouse point
+    // const distance = point.minus( vertex.positionProperty.value );
+    // if ( distance.magnitude !== 0 ) {
+    //
+    //   // const horizontalDirection = distance.x > 0 ? Vector2.X_UNIT.negated() : Vector2.X_UNIT;
+    //   // const horizontalRay = new Ray2( point, horizontalDirection );
+    //   //
+    //   // const verticalDirection = distance.y > 0 ? Vector2.Y_UNIT.negated() : Vector2.Y_UNIT;
+    //   // const verticalRay = new Ray2( point, verticalDirection );
+    //   //
+    //   const directDirection = vertex.positionProperty.value.minus( point ).normalized();
+    //   const directRay = new Ray2( point, directDirection );
+    //
+    //   // // get the intersection points
+    //   // const horizontalIntersections = vertex.dragAreaProperty.value!.intersection( horizontalRay );
+    //   // const verticalIntersections = vertex.dragAreaProperty.value!.intersection( verticalRay );
+    //   //
+    //   // if ( horizontalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( horizontalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.x = closestIntersection.point.x;
+    //   // }
+    //   // else if ( verticalIntersections.length > 0 ) {
+    //   //   const closestIntersection = _.minBy( verticalIntersections, intersection => intersection.point.distance( point ) );
+    //   //   closestPoint.y = closestIntersection.point.y;
+    //   // }
+    //
+    //   const directIntersections = vertex.dragAreaProperty.value!.intersection( directRay );
+    //   const closestIntersection = _.minBy( directIntersections, intersection => intersection.point.distance( point ) );
+    //   closestPoint.set( closestIntersection.point );
+    //
+    //   console.log( closestPoint, point );
+    // }
+
+    // return closestPoint;
+  }
+
   /**
    * Returns whether or not the quadrilateral shape is a parallelogram, within the tolerance defined by
    * angleToleranceIntervalProperty. This function uses parallelSidePairsProperty and requires that Property
@@ -1321,10 +1437,10 @@
    * Update the drag areas for all vertices.
    */
   private setVertexDragAreas(): void {
-    this.vertexA.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
-    this.vertexB.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
-    this.vertexC.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
-    this.vertexD.dragAreaProperty.set( this.createVertexArea( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
+    this.vertexA.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexA, this.vertexB, this.vertexC, this.vertexD ) );
+    this.vertexB.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexB, this.vertexC, this.vertexD, this.vertexA ) );
+    this.vertexC.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexC, this.vertexD, this.vertexA, this.vertexB ) );
+    this.vertexD.setDragAreaLines( this.createVertexDragAreaLines( this.model.modelBoundsProperty.value!, this.vertexD, this.vertexA, this.vertexB, this.vertexC ) );
   }
 
   public setPropertiesDeferred( deferred: boolean ): void {

```
</details>

The algorithm works but it continues to have the problem demonstrated in https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1111083130. Reducing the drag area in this way also doesn't work with the reduced area lines cross, for example here with lines CA and DA:
![image](https://user-images.githubusercontent.com/6396244/165993674-21d7746e-b143-4f53-947a-cfa800403434.png)

**Comment by jessegreenberg** [comment]
The behavior of dragging physical object in energy-forms-and-changes is quite similar to what we need here. Relevant code is in EFACIntroScreenView `constrainMovableElementMotion`.

**Comment by BLFiedler** [comment]
In model meeting, we decided we would go with the collision model changes in place of the shape boundaries created in #15 . This will make the model more "physical", which may not have worked if we wanted crossed shapes, but we are explicitly preventing them here. We chose this over trying to continue adding physical constraints to the geometrically based model JG had previously created.

**Comment by BLFiedler** [comment]
While other design decisions are being made, this should return to top priority to enable haptic feedback as well as sound at the boundaries.

**Comment by jessegreenberg** [comment]
During design meeting today @jbphet pointed out a case in faradays-law that also behaves pretty similarly to how we want it to behave here. Look for the collisions and dragging behavior between magnet and coils.

**Comment by jessegreenberg** [comment]
The `checkMotionAgainstObstacles` works by looking for overlap between the "leading edges" of the moving object and the "obstacles edges". The edges can only be horizontal or vertical lines and it looks like the implementation leverages that the edge lines are x or y axis aligned with no rotation. It uses that assumption when finding the intersection of edges and to determine the corrected translation for smooth dragging.

Here is the relevant part:
https://github.com/phetsims/faradays-law/blob/354b5533c2860e94b34305846f760c9d780189e6/js/faradays-law/model/FaradaysLawModel.js#L232-L263

If this part could be generalized to handle rotated bounds I think this is almost what we need.

**Comment by jessegreenberg** [comment]
I have been trying to use P2.js, a physics and collision detection library to support this instead of re-implementing collision detection for drag needs here. It has been a huge struggle with only moderate success. It still isn't working fully.

I found a different library called SAT (https://github.com/jriecken/sat-js). It is more light-weight but may be better for our needs here. It isn't a full physics engine but has collision detection support with examples for how to implement dragging like what we want in this sim. I am going to give that a try instead. I played around with it briefly and got as far as I got with P2.js in a few minutes. Id like to put P2.js on hold and continue with SAT for now.

**Comment by BLFiedler** [comment]
Note from Slack:

JG:
Back in the day we did some work to support various multitouch cases with quad, like how it should behave when dragging a side and vertex at the same time. Is that important to maintain? It might impact https://github.com/phetsims/quadrilateral/issues/41


BF:
Ah nice catch. It would be nice to support:
1)  at minimum any 2 vertices (4 ideal),
2) two opposite sides, e.g., side AB and side CD (I know adjacent sides can get weird) and
3) a side and it's opposite vertices (2-3 touch points, e.g, side AB, corner C and corner D) (edited)

**Comment by jessegreenberg** [comment]
Wahoo! Collision detection with SAT is working quite well.  We have a rigged quadrilateral model using SAT that corrects vertex/side/boundary overlap for rotated polygons and moving sides.

Side jitter will be corrected:

<details>

![Animation](https://user-images.githubusercontent.com/6396244/191386084-fe883d2f-e03b-4b8c-9655-0510e08a2338.gif)

</details>

In meeting today we observed that a vertex can "tunnel" through another object. A fix for that was to prevent the Vertex from moving too far in one step. But that makes the dragging feel a little sluggish. Its not a big deal, but there may be a better way to prevent this by not responding to overlap corrections if the overlap Vector flips over an axis. 

Ill keep thinking about that, and then get side dragging working with this model. Once that is in place we can put in the sim!

**Comment by jessegreenberg** [comment]
Ah, found the case where things really get messed up. If an object is sandwiched between more than one other and the collision detector is trying to move it into another object, others get pushed around to avoid new collisions.  But it is creating strange behavior like this:

<details>

![rts2](https://user-images.githubusercontent.com/6396244/196461753-d4dde330-c2e5-4f98-84e8-d413d1e0cfb4.gif)
</details>

**Comment by jessegreenberg** [comment]
11/29/22 meeting decision: We are out of time for this issue and we have decided to let it go. We can easily get rid of the stickiness around the boundary (done in https://github.com/phetsims/quadrilateral/commit/3a4f25a38738a6fa9cd24cca8c400c4639b9ff6a). But we are going to skip complex collision detection around the quadrilateral shape. Instead, we are just going to fix the "sticky case" bugs above, and document the approaches we tried in this issue. Ill add documentation in the following comment.

**Comment by jessegreenberg** [comment]
Here is a bit of a brain dump on things I tried and challenges:

The challenge of this issue was to detect a collision between two shapes AND find the most reasonable position for the dragging object as it moves along the intersecting segments.

I tried three different approaches to get this working.

1) Using p2.js (https://github.com/schteppe/p2.js):
- p2 is 2d physics engine with a lot of support for detecting shape collisions and applying responsive forces from overlap. In a first attempt, I use p2 to make a physical model of the quad where vertices and sides were p2 "bodies" with mass. I tired to use default p2 response to the overlap so that each animation frame bodies would be pushed out of eachother with modelled force. p2 supports physical "constraints" so that you can tie sides to connection points (vertices). I found it difficult to turn off parts of the physics while dragging objects. There were difficult cases where a vertex would get "pinched" between two sides and cause eratic behavior. The physical responses from constraints while using p2 had a "wobbly" feel. Using p2 (a fully fledged 2d physics engine) also seems very heavy for this collision detection problem. p2 also has no support for changing shapes, so it didn't work very well to model quadrilateral sides that grow/shrink in length.
- I tried to disable physics responses and use modified versions of p2 overlap detection/correction functions without force. In particular, correcting object positions by a "penetrationVector" of the ContactEquation: http://schteppe.github.io/p2.js/docs/classes/ContactEquation.html. p2 usually applies a force on the objects to push them in the directions of the penetrationVector but I tried to just move objects in the direction of that vector without applying forces for more controlled motion. See https://github.com/schteppe/p2.js/blob/9b282b5c3ce6369003d0ae29a2928f5456e01771/src/equations/ContactEquation.js#L101-L105 for where that is calculated in p2. I believe this should work but was not able to get it working well enough to support all quadrilateral motion in time.
2) Using SAT.js (https://github.com/jriecken/sat-js): 
- SAT.js is a collision detection library using the "separating axis theorem" (hence the name). It notifies when shapes overlap, and provides details of the overlap to correct the overlap. The library provides a "response" https://github.com/jriecken/sat-js#satresponse whenever there is a collision. It includes an overlap vector such that you can subtract the overlap vector from the object position to remove all overlap. That is great, but I had a hard time using it for the cases of quadrilateral. (For example, dragging a Vertex into a corner, (https://github.com/phetsims/quadrilateral/issues/41#issuecomment-1282503925, and dragging a vertex such that its connected side hits a different vertex).  Objects also tunnel through others if the mouse moves too quickly. p2 gets around this by using "ccd speed threshold": http://schteppe.github.io/p2.js/docs/files/src_objects_Body.js.html#l399.  SAT.js may provide everything necessary to solve this problem, but I could not successfully use the responses to handle all overlap cases needed by quad.
3) Using a custom solution (a la faradays-law)
Faradays Law has a similar requirement for preventing overlap and I tried to borrow aspects of that implementation and use over here. See https://github.com/phetsims/faradays-law/blob/354b5533c2860e94b34305846f760c9d780189e6/js/faradays-law/model/FaradaysLawModel.js#L404-L413. That solution assumes all objects are axis-aligned rectangles so we worked to generalize it. We came up with a similar "leading edge" solution where segments extended from the leading edges to the proposed position could be used to calculate overlap. A projection (from a dot product) along the constraining edge could determine best alternative translation without overlap. That way an object could "slide" along a boundary with available x/y space without simply stopping at the intersection point.
![image](https://user-images.githubusercontent.com/6396244/204915505-54c6edaa-6803-4b7b-9600-bf11b93e6881.png)

Theoretically this might work! But there are many cases I didn't get to supporting such as handling overlap with multiple objects at a time. This method also doesn't have a solution for when the end points of a leading edge extend outside of the boundary edge (for example, moving a side that is longer than a vertex into a vertex, either by dragging directly or rotating a side by dragging a Vertex. Also note that to prototype this I had to modify `kite.Line.intersectOther`.  For some reason, that intersection function does not count an intersection if a segment point rests exactly on a line, which caused problems in this case (if a point on the test segment rests exactly on a boundary segment, it should be considered an intersection and corrected). 

OK, closing this issue for now. Maybe some day we will return to it.
